src/V3ThreadPool.h

```diff
@@ -0,0 +1,174 @@
+    // and resumes execution of other jobs.
+    void request_exclusive_access(job_t exclusive_access_job) VL_MT_SAFE {
+        if (executeSynchronously()) {
+            exclusive_access_job();
+        } else {
+            VerilatedLockGuard stoppedJobLock(m_stoppedJobsMutex);
+            m_stoppedJobs++;
+            m_cv.notify_all();
+            m_stoppedJobsCV.wait(stoppedJobLock, [&]() VL_REQUIRES(m_stoppedJobsMutex) {
+                return m_stoppedJobs == m_workers.size();
+            });
+            exclusive_access_job();
+            m_stoppedJobs = 0;
+            m_stoppedJobsCV.notify_all();
+        }
+    }
+
+    // Similar to above function, but when function that requires exclusive access is behind mutex,
+    // this function can also unlock it while it is waiting for other jobs and lock afterwards.
+    // It can't be merged with above function due to limitations in clang thread-safety analysis.
+    void request_exclusive_access(VerilatedMutex* locking, job_t exclusive_access_job) VL_MT_SAFE
+        VL_REQUIRES(locking) {
+        if (executeSynchronously()) {
+            exclusive_access_job();
+        } else {
+            VerilatedLockGuard stoppedJobLock(m_stoppedJobsMutex);
+            locking->unlock();
+            m_stoppedJobs++;
+            m_cv.notify_all();
+            m_stoppedJobsCV.wait(stoppedJobLock, [&]() VL_REQUIRES(m_stoppedJobsMutex) {
+                return m_stoppedJobs == m_workers.size();
+            });
+            locking->lock();
+            exclusive_access_job();
+            m_stoppedJobs = 0;
+            m_stoppedJobsCV.notify_all();
+        }
+    }
+
+    // This function can be used to check if other thread requested exclusive access to processing,
+    // if so, it waits for it to complete. Afterwards it is resumed.
+    void check_stop_requested() VL_MT_SAFE {
+        VerilatedLockGuard stoppedJobLock(m_stoppedJobsMutex);
+        if (m_shutdown || !m_stoppedJobs) return;
+        m_stoppedJobs++;
+        m_stoppedJobsCV.notify_all();
+        m_stoppedJobsCV.wait(stoppedJobLock,
+                             [&]() VL_REQUIRES(m_stoppedJobsMutex) { return m_stoppedJobs == 0; });
+    }
+
+    // Similar to above function, but when we checking function is behind mutex,
+    // this function can also unlock it while it waits to be resumed.
+    void check_stop_requested(VerilatedMutex* locking) VL_MT_SAFE VL_REQUIRES(locking) {
+        VerilatedLockGuard stoppedJobLock(m_stoppedJobsMutex);
+        if (m_shutdown || !m_stoppedJobs) return;
+        locking->unlock();
+        m_stoppedJobs++;
+        m_stoppedJobsCV.notify_all();
+        m_stoppedJobsCV.wait(stoppedJobLock,
+                             [&]() VL_REQUIRES(m_stoppedJobsMutex) { return m_stoppedJobs == 0; });
+        locking->lock();
+    }
+
+private:
+    bool executeSynchronously() const VL_MT_SAFE { return m_workers.empty(); }
```

Comment:
```suggestion
    bool willExecuteSynchronously() const VL_MT_SAFE { return m_workers.empty(); }
```

The name otherwise suggests it's going to execute something.